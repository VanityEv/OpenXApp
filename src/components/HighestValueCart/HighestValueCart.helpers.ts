import { Cart, Product, ProductIdentifier, User } from "../../APIData.types";
import { CartWithOwner } from "./HighestValueCart.types";

export function getHighestValueCart(
  cartsData: Cart[],
  productsData: Product[],
  usersData: User[]
) {
  const productsMap = getProductsMap(productsData);

  let result: CartWithOwner = {
    value: 0,
    cartID: 0,
    owner: {
      firstname: "",
      lastname: "",
    },
  };

  let currentValue = 0;
  let productPrice;
  cartsData.forEach((cart: Cart) => {
    currentValue = 0;
    cart.products.forEach((product: ProductIdentifier) => {
      productPrice = productsMap.get(product.productId);
      if (productPrice !== undefined) {
        currentValue += productPrice * product.quantity;
      }
    });

    if (currentValue > result.value) {
      result.value = currentValue;
      result.cartID = cart.id;
      result.owner = usersData[cart.userId - 1].name;
    }
  });
  return result;
}

function getProductsMap(data: Product[]) {
  const productsMap = new Map<number, number>();
  data.forEach((product: Product) => {
    if (!productsMap.has(product.id)) {
      productsMap.set(product.id, product.price);
    }
  });
  return productsMap;
}
