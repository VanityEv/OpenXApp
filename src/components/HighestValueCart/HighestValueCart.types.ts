import { Name } from "../../APIData.types";

export type CartWithOwner = {
  value: number;
  cartID: number;
  owner: Name;
};
