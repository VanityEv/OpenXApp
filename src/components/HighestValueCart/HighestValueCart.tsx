import { useEffect, useState } from "react";
import { Cart, Product, User } from "../../APIData.types";
import { CartWithOwner } from "./HighestValueCart.types";
import { getHighestValueCart } from "./HighestValueCart.helpers";
import { Box, Paper, Typography } from "@mui/material";

type HighestValueCartProps = {
  cartsData: Cart[];
  productsData: Product[];
  usersData: User[];
};

export const HighestValueCart = ({
  cartsData,
  productsData,
  usersData,
}: HighestValueCartProps) => {
  const [highestValueCart, setHighestValueCart] = useState<CartWithOwner>();
  useEffect(() => {
    setHighestValueCart(
      getHighestValueCart(cartsData, productsData, usersData)
    );
  }, [cartsData, productsData, usersData]);
  return (
    <Paper elevation={3} sx={{ p: 3 }}>
      <Typography variant="h5" sx={{ textAlign: "center" }}>
        Task 3
      </Typography>
      <Typography variant="subtitle1" sx={{ textAlign: "center" }}>
        Highest Value Cart:
      </Typography>
      <Box>
        <Typography sx={{ textAlign: "center" }} display="inline">
          Owner:
        </Typography>
        <Typography display="inline" sx={{ fontWeight: "bold" }}>
          {highestValueCart?.owner.firstname +
            " " +
            highestValueCart?.owner.lastname}
        </Typography>
      </Box>
      <Box>
        <Typography sx={{ textAlign: "center" }} display="inline">
          Value:
        </Typography>
        <Typography sx={{ fontWeight: "bold" }} display="inline">
          {highestValueCart?.value.toFixed(2)}
        </Typography>
      </Box>
    </Paper>
  );
};
