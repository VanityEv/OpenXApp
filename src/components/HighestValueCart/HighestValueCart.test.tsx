import { render, screen } from "@testing-library/react";
import { HighestValueCart } from "./HighestValueCart";
import {
  mockCart,
  mockProducts,
  mockUser,
} from "../../test/testData.constants";

describe("Component: HighestValueCart", () => {
  it("should render name of the owner with the highest value cart", () => {
    render(
      <HighestValueCart
        cartsData={mockCart}
        productsData={mockProducts}
        usersData={mockUser}
      />
    );

    const owner = screen.getByText("danny doe");

    expect(owner).toBeInTheDocument();
  });
  it("should render value of the cart", () => {
    render(
      <HighestValueCart
        cartsData={mockCart}
        productsData={mockProducts}
        usersData={mockUser}
      />
    );

    const value = screen.getByText("1279.48");
    expect(value).toBeInTheDocument();
  });
  it("should render the same component", () => {
    const { container } = render(
      <HighestValueCart
        cartsData={mockCart}
        productsData={mockProducts}
        usersData={mockUser}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
