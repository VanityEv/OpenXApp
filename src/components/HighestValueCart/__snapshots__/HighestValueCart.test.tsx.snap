// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: HighestValueCart should render the same component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-1yocq7i-MuiPaper-root"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 css-ph2xsj-MuiTypography-root"
    >
      Task 3
    </h5>
    <h6
      class="MuiTypography-root MuiTypography-subtitle1 css-118o8hd-MuiTypography-root"
    >
      Highest Value Cart:
    </h6>
    <div
      class="MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-llyu5n-MuiTypography-root"
      >
        Owner:
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1c1be42-MuiTypography-root"
      >
        danny doe
      </p>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-llyu5n-MuiTypography-root"
      >
        Value:
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1c1be42-MuiTypography-root"
      >
        1279.48
      </p>
    </div>
  </div>
</div>
`;
